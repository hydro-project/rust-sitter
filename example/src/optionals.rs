#[rust_sitter::grammar("optionals")]
#[allow(dead_code)]
mod grammar {
    use rust_sitter::Spanned;

    #[rust_sitter::language]
    #[derive(Debug)]
    pub struct Language {
        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
        v: Option<i32>,
        #[rust_sitter::leaf(text = "_")]
        _s: (),
        t: Spanned<Option<Number>>,
        #[rust_sitter::leaf(text = ".")]
        _d: Option<()>,
    }

    #[derive(Debug)]
    pub struct Number {
        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
        v: i32,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn optional_grammar() {
        insta::assert_debug_snapshot!(grammar::parse("_"));
        insta::assert_debug_snapshot!(grammar::parse("_."));
        insta::assert_debug_snapshot!(grammar::parse("1_"));
        insta::assert_debug_snapshot!(grammar::parse("1_."));
        insta::assert_debug_snapshot!(grammar::parse("1_2"));
        insta::assert_debug_snapshot!(grammar::parse("1_2."));
        insta::assert_debug_snapshot!(grammar::parse("_2"));
        insta::assert_debug_snapshot!(grammar::parse("_2."));
    }
}
