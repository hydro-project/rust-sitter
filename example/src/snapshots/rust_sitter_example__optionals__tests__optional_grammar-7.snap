---
source: example/src/optionals.rs
expression: "grammar::parse(\"_2\")"
---
Ok(
    Language {
        v: None,
        _s: (),
        t: Spanned {
            value: Some(
                Number {
                    v: 2,
                },
            ),
            span: (
                1,
                2,
            ),
        },
        _d: None,
    },
)
